man/dbcgen.1
.TH DBCGEN 1 "12 September 1995"
.FB
.SH NAME
dbcgen \- generate a converted \fIFirstBase\fP database
.SH SYNOPSIS
.B dbcgen
[\fB-b\fP] [\fB-y\fP] [\fB-B\fP]
[\fB-d\fP \fIdatabase\fP] [\fB-i\fP \fIindex\fP]
.SH DESCRIPTION
.I Dbcgen
generates a converted \fIFirstBase\fP database using a conversion 
dictionary defined via
\fIdbdcnv\fP(1).
This conversion dictionary is used to determine 
the size, shape, and structure of the new database.
.PP
This tool uses a \fIFirstBase\fP index to generate the output in sorted order.
If this index is detected to be out of date or non-existent, it will
be regenerated \fIautomatically\fP. An index is out of date if 
either the database or the index dictionary have changed.
.PP
The \fB-b\fP flag will run the \fIFirstBase\fP tool in batch mode,
suitable for shell scripts.
This mode produces no output on the terminal, thus can also be used
when running the process in the background.
The \fB-y\fP
flag will run the tool at the terminal without pausing to ask
the user if existing files can be destroyed and without displaying
a 'picture' of the requested task.
.PP
The \fB-B\fP flag will enable the fastest throughput of
records from the database. In this case, the entire database is used,
and the \fIdbcgen\fP generation is speeded up by internal record buffering.
Note that an index will be NOT be used with record buffering,
and record locking is disabled. However, the \fB-i\fP argument is still
required so that \fIdbcgen\fP can locate the
the conversion dictionary (\fIdbdcnv\fP(1)).
.PP
The following discussion refers to using the generator interactively
without the batch mode or yes mode flags.
.PP
\fIDbcgen\fP(1)
first displays an image of the converted database-to-be and
its file name.
The user has a chance to abort here. After the database conversion, 
\fIdbcgen\fP(1) creates a new database dictionary for the new database.
.PP
Dbcgen can change the size, shape and structure of a valid \fIFirstBase\fP 
database. Existing database fields can be lengthened, shortened, or deleted.
New fields can also be added.
See 
\fIdbdcnv\fP(1)
for more conversion details.
.PP
Note: \fIdbclean\fP(1)
is automatically run at the completion of this program to
create a map file for the new database.
.SH FILES
.PD 0
.TP 10
dbase
default name for the \fIFirstBase\fP database.
.TP 10
index
default name for the \fIFirstBase\fP index.
.TP 10
*.?dict
dictionary for \fIFirstBase\fP dbase or index.
.TP 10
*.idictc
dictionary for conversion generator.
.TP 10
*.cdb
\fIFirstBase\fP database.
.TP 10
*.map
record map of \fIFirstBase\fP database.
.TP 10
*.idx
\fIFirstBase\fP index.
.PD
.SH SEE ALSO
dbdbas(1), dbedit(1), dbdcnv(1), generators(5)
.PP
.I FirstBase User's Guide and Reference Manual
.br
.SH DIAGNOSTICS
\fIDbcgen\fP(1)
pauses to ask permission to destroy an existing
\fIFirstBase\fP database. A valid \fIFirstBase\fP database dictionary is
required for the input database.
.br


ChangeLog
2012-04-14    <john.bolding@epona.horseweb.com>

	* modified libscr/input.c -- read_kcheck -- to avoid a problem when:
	     EDITINPUT is OFF
	     two (2) ESCAPE's were hanging in read_kcheck

2011-09-29    <john.bolding@epona.horseweb.com>

	* added fflush(-1) to dbmacro via libmexpr/mac_file.c

2009-02-25    <john.bolding@epona.horseweb.com>

	* changed src/dbsql/tokenscript to allow for space between # and define.

2006-09-19  John  <john@cerebus.horseweb.com>

	* complete rewrite of libcc/subline.c to NOT write on the passed in s.

2003-03-29  John Bolding  <john@firstbase.com>

        * tweaked libmpars/tokenscript and dbsql/tokenscript
          to accomodate the bison output "# define" instead of "#define".

2001-09-01  John Bolding  <john@firstbase.com>

        * tweaked configure.in to not use termios.h if OS is sco3.2
	* src/libmexpr/mac_func.c
	* src/libmexpr/mac_str.c
	  augmented dbmacro to include
	     crypt(string, salt), returns an encrypted string ala crypt(3)

2001-03-19  John Bolding  <john@firstbase.com>

	* src/libmexpr/mac_sec.c: 
	* src/libmexpr/mac_func.c: 
	* src/libmexpr/mac_file.c: 
	* src/include/macro.h:
	  modified dbmacro functions as follows:
	  - modified mf_creat() to do a chmod on the file if it already exists.
	  - added an mf_mkstemp() that really calls mf_mktemp()
	    (since mf_mktemp really uses mkstemp(2), it closes the fd.
	     a real mf_mkstemp would modify the array passed in for
	     the user to keep. that did not happen this go around)
	  - renamed all the firstbase secure record functions to 
            r_owner() r_group() r_mode() r_chown() r_chgrp() r_chmod()
	  - added an mf_chmod() command

2001-03-18  John Bolding  <john@firstbase.com>

	* src/dbtools/dbmacro.c: fixed path_trans to be null before
	  its used. fixed to only look at extension of file name, .cgi,
	  to determine whether to go cgi mode. this is so that i can
	  test with program names like ndbmacro.cgi.

	* src/libcdb/cgi.c: modified cgi_read_get to FB_ERROR out
	  if the QUERY_STRING env variable is not set.

	* src/libcdb/System.c: seems linux system() call is a bit
	  flaky at times on some versions, so fb_system() has been
	  modified to not use the system() call at all.
	  now it uses execvp. in addition, changed to waitpid()
	  to try and get better response from cgi dbmacro scripts that
	  use the macro call system(). and, now pays attention to
	  status from waitpid() and exits if its strange.

2001-02-23  John Bolding  <john@firstbase.com>

	* src/libscr/screen.c: modified visible bell area to
	  use a ^G if its not defined.
	  this was a terminfo only bug --- termcap was already working.

2001-02-17  John Bolding  <john@firstbase.com>

	* release 9.3.1
	* man pages for section 4 were missing from distribution.
	  (finally noticed by <garak_dracul@hotmail.com>).

2001-02-16  John Bolding  <john@firstbase.com>

	* man.ps target now in man/Makefile

	* fixed use of mktemp and gets system wide

	* src/libinit/setup.c:
	  changed default behaviour. these are now default:
	     MENUPAUSE OFF
	     MAKEINDEX OFF
	     DECIMAL ON

	* quite a few bugs regarding #if clauses and FSYNC/FCNTL
	  these should have been HAVE_FSYNC or HAVE_FCNTL.
	  some were wrong. some were using FSYNC instead of HAVE_FSYNC.
	  this might have caused multiuser problems with records
	  stored across slow devices (NFS, etc).
	  all of these were introduced when firstbase went autoconf.
	
	* src/libcdb/System.c:
	  - added \" to list of characters that cause a complete system() call
	  - make return status better for those that fork and wait

	* a number of little bugs concering the full screen display
	  for devices bigger than 24 x 80.
	
	* 180 degree about face regarding TERMCAP vs TERMINFO:
	  since ncurses seems to be pretty common on Linux/FreeBSD/Etc,
	  the default is now to use the ncurses library and if this
	  is not found the HAVE_LIBNCURSES is undef'ed in
	  src/include/config.h.

	  Although I am not a TERMINFO fan, perhaps this will be good.
	  At some point, the firstbase screen libraries could
	  make better use of ncurses to optomize screen motion, etc.
	  Perhaps even support for colors, pop-up style ncurses items, etc.

	  So, TERMINFO via ncurses is the default.
	  And, TERMCAP is done only if necessary.

2001-02-05  John Bolding  <john@firstbase.com>

	* src/libscr/input.c: 
	* src/libscr/e_input.c: both of these were modified to handle
	  windows that change sizes while being used, ala SIGWINCH.

	* src/libscr/refresh.c: 
	* src/libscr/scrhdr.c: 
	* src/libscr/scrstat.c: 
	* src/libscr/scrtime.c: 
	* src/libinit/ender.c: 
	* src/libcdb/screrr.c:
	  all of these were modified to better handle standard
	  screen messages on surfaces bigger than 24 x 80,
	  and to restore headers and footers when the frame size changes.

	* src/libcdb/winsize.c: modified to handle different frame/window
	  sizes using an ioctl() call with TIOCGWINSZ or TIOCGSIZE,
	  then allowing environment variables LINES and COLUMNS to
	  override all else. (Thanks to <vanonim@dial.eunet.ch>).

	  finally, sanity checks are done to make sure
	  nothing interactive less than 24 x 80 gets through.

	* src/libinit/getargs.c: took out code that was getting in the way
	  of the new fb_winsize(). all sanity checks are now done at
	  screen setup time.

	* src/libscr/screen.c: force all columns and lines determinations
	  to be done inside of winsize, newly named fb_winsize().

2001-02-02  John Bolding  <john@firstbase.com>

	* src/libinit/setup.c: removed the LINES setup variable,
	     (also force makeindex to be OFF at startup).

2001-02-02  John Bolding  <john@firstbase.com>

	* .visual: changed to /usr/bin/vi (was /usr/ucb/vi).

	* configure.in (INCLUDES): made `ed' and one of `flex' or `lex'
	  and one of `bison', `byacc', or `yacc' required in order
	  to complete ./configure.

$Id: ChangeLog,v 9.9 2006/09/19 17:46:07 john Exp $


src/dbpgen/dbcgen.c
/*
 * Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.
 *
 * $Id: dbcgen.c,v 9.1 2001/02/16 19:44:55 john Exp $
 *
 * See the file LICENSE for conditions of use and distribution.
 *
 */

#ifndef LINT
static char sccsid[] = "@(#)dbcgen.c	8.1 8/19/93 FB";
#endif

/* 
 *  dbcgen - conversion generator. convert a Cdb fb_database to
 *    a newly structured dbase with image/expanded/truncated fields.
 */
 
#include <fb.h>
#include <fb_vars.h>

/*
 *  assumption here is that no more than FB_MAXKEEP fields will be converted
 *  at any one time. this is needed since any number of new fields can be 
 *  used...perhaps i just got lazy. hopefully, 1000 is enough.
 *
 *  keyprint is the list of new fields.
 *  csize (matches one for one). its value determines the
 *     status of the converted fb_field. < 0 = new.
 *     positive sizes are compared to determine whether 
 *     old fb_field is image/truncated/expanded
 */

extern short int cdb_secure;
extern char *cdb_DBCLEAN;
extern char *cdb_S_EOREC;

static char *FMT1 = "%04d%03d%s";		/* secure bits */
char perm[FB_MAXNAME];

static fb_field *keyprint[FB_MAXKEEP] = { NULL };
static fb_field **pp;
static int csize[FB_MAXKEEP] = { 0 };
fb_database *hp;
static char *fld;
static long wcount;				/* write counter */
static Bflag = 0;				/* Blocking flag */

static cgen();
static condict();

/*  
 *  dbcgen main driver
 */
 
   main(argc, argv)
      int argc;
      char *argv[];
   
      {
         char title[FB_MAXLINE], filen[FB_MAXNAME], source[FB_MAXNAME], 
	      dest[FB_MAXNAME], msg[FB_MAXLINE];
         FILE *initprint();
	 int fd;

         fb_getargs(argc, argv, FB_ALLOCDB);
	 hp = cdb_db;
	 if (fb_testargs(argc, argv, "-B") > 0)
	    Bflag = 1;
	 fb_opendb(cdb_db, READ, FB_WITHINDEX, FB_MUST_INDEX);
	 fb_scrhdr(cdb_db, "Parsing"), fb_infoline();
         initprint(cdb_keymap, keyprint, hp, title, filen,
	       hp->idict, (int *) 0, csize);
	 close(creat(filen, 0666));		
	 fd = fb_mustopen(filen, 2);
	 fb_rootname(dest, filen);
	 fb_rootname(source, hp->dbase);
	 if (equal(source, dest))
	    fb_xerror(FB_MESSAGE, SYSMSG[S_ILLEGAL], "same files");
	 if (!cdb_batchmode && !cdb_yesmode){
            ctrace(keyprint, filen, csize);
	    if (fb_mustbe('y',"If accurate, enter 'y' to continue: ",
	          cdb_t_lines, 1) == FB_ERROR)
               fb_xerror(FB_ABORT_ERROR, SYSMSG[S_NOT_DONE], NIL);
	    fb_move(4, 1); fb_clrtobot(); fb_infoline();
	    }
         pp = keyprint;
         if (cgen(fd, filen) == FB_ERROR)
            fb_xerror(FB_ABORT_ERROR, SYSMSG[S_NOT_DONE], NIL);
	 fb_closedb(cdb_db);
	 if (!cdb_batchmode){
	    fb_scrstat("Cleaning");
	    fb_move(4,1), fb_clrtobot();
	    sprintf(msg, "Cleaning %s", dest);
	    fb_fmessage(msg);
	    }
	 sprintf(title, "%s -b %s", cdb_DBCLEAN, dest); /* borrow title */
	 fb_system(title, FB_WITHROOT);
	 fb_ender();
      }
      
/* 
 *  cgen - generate converted data base according to list in pp 
 */
 
   static cgen(fd, filen)
      char filen[];
      int fd;
      
      {
         char fname[FB_MAXNAME];
         int gen_one();

         wcount = 0L;
         if (!cdb_batchmode){
            fb_scrstat("Converting");
	    FB_XWAIT();
	    fb_gcounter(wcount);
	    }
	 if (cdb_yesmode)
	    sleep(1);				/* some are too fast */
         fld = cdb_afld;				/* set local to global */

	 fb_putseq(fd);
	 fb_puthead(fd, 0L, 0L);
         if (cdb_secure){
            if (fb_putmode(fd, fb_getuid(), fb_getgid(), "666") == FB_ERROR)
               fb_xerror(FB_IO_ERROR, filen, NIL);
	    sprintf(perm, FMT1, fb_getuid(), fb_getgid(), "666");
            }
         fb_w_init(1, fd, -1);
         if (Bflag == 0)
	    fb_forxeach(hp, gen_one);
         else
	    fb_blockeach(hp, gen_one);
         fb_wflush(1);
         if (fb_puthead(fd, wcount, 0L) == FB_ERROR)
	    fb_xerror(FB_IO_ERROR, SYSMSG[S_BAD_HEADER], hp->dbase);
	 close(fd);
         fb_basename(fname, filen);
         strcat(fname, ".ddict");
         condict(fname);
	 return(FB_AOK);
      }

/*
 * gen_one - generate exactly one -- called from a foreach style loop
 */

   gen_one(hp)
      fb_database *hp;

      {
         int k, nc;

         for (k = 0; (pp[k]->size) != 0; k++){
            if (pp[k]->type != FB_FORMULA && pp[k]->dflink == NULL){
               fld[0] = NULL;
               if (csize[k] >= 0){
                  strcpy(fld, pp[k]->fld); /* used to allow truncation */
                  if (pp[k]->type != FB_BINARY){
                     if (csize[k] < pp[k]->size)
                        fld[csize[k]] = NULL;
                     fb_trim(fld);
                     }
                  }
               else{
                  if (pp[k]->type == FB_BINARY)
                     for (nc = pp[k]->size; nc > 0; nc--)
                        fb_w_write(0, "\000");
                  }
               if (pp[k]->type != FB_BINARY){
                  fb_nextwrite(0, fld);
                  fb_w_write(0, "\000");
                  }
               else if (csize[k] >= 0)
                  fb_w_writen(0, fld, pp[k]->size);
               }
            }
         fb_nextwrite(0, " ");      	/* deletion place holder */
         if (cdb_secure)
            fb_nextwrite(0, perm);      /* permissions */
         fb_w_write(0, "\000");
         fb_w_write(0, cdb_S_EOREC);	/* end of record marker */
         ++wcount;
         if (!cdb_batchmode){
            fb_gcounter(wcount);
            }
      }

/* 
 *  condict - output the dict to the converted dbase dictionary file
 */
 
   static condict(f)
      char *f;
      
      {
         int i;
         FILE *fs, *fb_mustfopen();
   
         fs = fb_mustfopen(f, "w");
         for (i = 0; pp[i]->size != 0; i++){
            fb_unders(pp[i]);
	    if (csize[i] < 0)
	       csize[i] = -csize[i];
	    pp[i]->size = csize[i];
	    fb_sdict(pp[i], fs, 0);
            }
         fclose(fs);
      }
   
/* usage message */
   usage()
      {
         fb_xerror(FB_MESSAGE,"usage: dbcgen [-b] [-y] [-d dbase] [-i index]", NIL);
      }
